naptime-graphql/src/main/scala/org/coursera/naptime/ari/graphql/SangriaGraphQlSchemaBuilder.scala

```diff
@@ -56,219 +39,39 @@ class SangriaGraphQlSchemaBuilder(
 
   /**
     * Generates a GraphQL schema for the provided set of resources to this class
-    * Returns a "root" object that has one field available for each Naptime Resource provided.*
+    * Returns a "root" object that has one field available for each Naptime Resource provided.
     *
-    * @return a Sangria GraphQL Schema with all resources defined
+    * @return a Sangria GraphQL Schema with all resources defined,
+    *         and a list of errors that came up while generating the schema
     */
-  def generateSchema(): Schema[SangriaGraphQlContext, DataMap] = {
-    val topLevelResourceObjects = for {
-      resource <- resources
-      resourceObject <- (try {
-        val resourceName = ResourceName(
-          resource.name, resource.version.getOrElse(0L).toInt).identifier
-        generateLookupTypeForResource(resourceName)
-      } catch {
-        case e: Throwable => None
-      }).toList if resourceObject.fields.nonEmpty
-    } yield {
-      Field.apply[SangriaGraphQlContext, DataMap, DataMap, Any](
-        formatResourceTopLevelName(resource),
-        resourceObject,
-        resolve = (context: Context[SangriaGraphQlContext, DataMap]) => {
-          Value(new DataMap())
-        })
+  def generateSchema(): WithSchemaErrors[Schema[SangriaGraphQlContext, DataMap]] = {
+    val topLevelResourceObjectsAndErrors = resources.map { resource =>
+      val lookupTypeAndErrors =
+        NaptimeTopLevelResourceField.generateLookupTypeForResource(resource, schemaMetadata)
```

Comment:
I think you're asking why I'm choosing to making the `generateLookupTypeForResource` static (and put it on an object instead of class). I think there are a few reasons / benefits for doing it this way:
- Separation of concerns: The field generator shouldn't have to maintain its own state of the world. If each field-builder injected in its own schemaMetadata, there's a lot more "state" in the application, and it would be much easier for something to get out of sync or for some oversight while coding. Right now, the SchemaProvider is a singleton and only has one copy of state. If it wasn't a singleton, or had some sort of side effect when loading the schema, or wasn't fully atomic, then different field-builders could technically have different schemas.
- Functional programming: I actually find it much easier to test an individual static function that takes in all of the state it needs, rather than having state part of a class that needs to be initialized. Mocking should work the same way for both classes and functions, but there's less state to worry about. (And no risk of side effects if you call the function multiple times)
- Static objects can be used from other static objects / functions. If the parent calling this was an object, and this was a class, we'd have to convert the parent into a class and inject this in, which could add further complexity.

Let me know if any of this doesn't make sense -- happy to chat more in person :)